qiskit/result/counts.py

```diff
@@ -0,0 +1,83 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A container class for counts from a circuit execution."""
+
+from qiskit.result import postprocess
+from qiskit.result import exceptions
+
+
+class Counts(dict):
```

Comment:
I worry about inheriting directly from `dict`. This is a notoriously tricky problem in Python as its backed by a C implementation. Perhaps its better to implement `MutableMapping`?

Reply:
In my experience it's not an issue as long as you don't override the object model methods dict relies on like `__getitem__`, `__setitem__` etc. The problems arise because the c impl doesn't actually use those methods for builtins like `update()`. In the case of this object it should be fine because this is just a dict with some helper methods added and extra attributes which won't conflict with dict in cpython. Using dict has advantage for backwards compat especially in #4485 where we start using this instead of a dict since it can be dropped in anywhere with minimal boiler plate. `MutableMapping` would work fine too but would require implementing all the object model methods, which would likely just be one liners calling an inner dict attribute.

Reply:
Yes, I figured this is being done this way for isinstance checks 👍. I agree with you about where the problems come from. The difficulty is in conveying this information to future developers I think.

Reply:
I can add a comment here in the next rev to try and explain this, or at least say we shouldn't override the dunder methods.

Reply:
Sounds good 👍 .

Reply:
Added in: https://github.com/Qiskit/qiskit-terra/pull/4501/commits/05252304cc110fa14b8d7ec96d155515fa3d36c2